### ISSUE TITLE
Publish the Model code for non-mnist tasks

### ISSUE THREAD

**Issue by NHDaly** [openai/InfoGAN#9]
The code in `models/regularized_gan.py` ([here](https://github.com/openai/InfoGAN/blob/master/infogan/models/regularized_gan.py#L30)) defines the tensorflow model for training a GAN network to generate _mnist_ data.

The `if network_type == "mnist":` line strongly implies that the model would look different for other tasks.

Can this same model be used for the other tasks described in the paper, such as 3d faces or CelebA? (Assuming we change the input codes, of course.)

Thanks! :)

**Comment by zdx3578** [comment]
in paper C Experiment Setup show just arch 


CelebA
The network architectures are shown in Table 3. The discriminator D and the recognition network Q shares most of the network. For this task, we use 10 ten-dimensional categorical code and 128 noise variables, resulting in a concatenated dimension of 228.
Table 3: The discriminator and generator CNNs used for SVHN dataset.
discriminator D / recognition network Q generator G
Input 32 ⇥ 32 Color image     Input 2 R228
4 ⇥ 4 conv. 64 lRELU. stride 2       FC. 2 ⇥ 2 ⇥ 448 RELU. batchnorm
4 ⇥ 4 conv. 128 lRELU. stride 2. batchnorm 4 ⇥ 4 upconv. 256 RELU. stride 2. batchnorm 4 ⇥ 4 conv. 256 lRELU. stride 2. batchnorm 4 ⇥ 4 upconv. 128 RELU. stride 2.
FC. output layer for D,
FC.128-batchnorm-lRELU-FC.output for Q 4 ⇥ 4 upconv. 64 RELU. stride 2.
4 ⇥ 4 upconv. 3 Tanh. stride 2.

**Comment by NHDaly** [comment]
Ah yes, thanks! I actually found that myself later. I should've posted it. thanks for the find! :)

That said, it'd be nice to know if my reimplementation of the above is correct... Thanks! :)

**Comment by zdx3578** [comment]
but i think 
latent_spec = [
        (Uniform(62), False),
        (Categorical(10), True),
        (Uniform(1, fix_std=True), True),
        (Uniform(1, fix_std=True), True),
    ]

this is not enough,
in paper:
MNIST, we choose to model the latent codes with one categorical code, c1 ⇠ Cat(K = 10, p = 0.1), which can model discontinuous variation in data, and two continuous codes that can capture variations that are continuous in nature: c2 , c3 ⇠ Unif ( 1, 1).

but what to express:   Street View House Number (SVHN
we make use of four 10 dimensional categorical variables and two uniform continuous variables as latent codes.

CelebA
In this dataset, we model the latent variation as 10 uniform categorical variables, each of dimension 10.

**Comment by zdx3578** [comment]
if isinstance(dist, Gaussian):
                assert dist.dim == 1, "Only dim=1 is currently supported"
                c_vals = []
                for idx in xrange(10):
                    c_vals.extend([-1.0 + idx * 2.0 / 9] * 10)
                c_vals.extend([0.] * (self.batch_size - 100))
                vary_cat = np.asarray(c_vals, dtype=np.float32).reshape((-1, 1))
                cur_cat = np.copy(fixed_cat)
                cur_cat[:, offset:offset+1] = vary_cat
                offset += 1
            elif isinstance(dist, Categorical):
                lookup = np.eye(dist.dim, dtype=np.float32)
                cat_ids = []
                for idx in xrange(10):
                    cat_ids.extend([idx] * 10)
                cat_ids.extend([0] * (self.batch_size - 100))
                cur_cat = np.copy(fixed_cat)
                cur_cat[:, offset:offset+dist.dim] = lookup[cat_ids]
                offset += dist.dim
            elif isinstance(dist, Bernoulli):
                assert dist.dim == 1, "Only dim=1 is currently supported"

**Comment by zdx3578** [comment]
embedding_dim = 100

latent_spec = [
    (Uniform(64), False),
    (Categorical(32), True),
]
con_latent_spec = [
    (LatentGaussian(embedding_dim), True)
]
https://github.com/RutgersHan/InfoGAN/blob/dev_auto/launchers/generate_images.py

good!

**Comment by zdx3578** [comment]
C.3 CelebA
The network architectures are shown in Table 3. The discriminator D and the recognition network Q shares most of the network. For this task, we use 10 ten-dimensional categorical code and 128 noise variables, resulting in a concatenated dimension of 228.

is

latent_spec = [
    (Uniform(128), False),
    (Categorical(10), True),
    (Categorical(10), True),
    (Categorical(10), True),
    (Categorical(10), True),
    (Categorical(10), True),
    (Categorical(10), True),
    (Categorical(10), True),
    (Categorical(10), True),
    (Categorical(10), True),
    (Categorical(10), True),
]
but how this??

C.5 Chairs
The network architectures are shown in Table 6. The discriminator D and the recognition network Q shares the same network, and only have separate output units at the last layer. For this task, we use 1 continuous latent code, 3 discrete latent codes (each with dimension 20), and 128 noise variables, so the input to the generator has dimension 189.

**Comment by NHDaly** [comment]
The above latent_spec worked okay for me.

```python
c3_celebA_latent_spec = [
    (Uniform(128), False),  # Noise
    (Categorical(10), True),
    (Categorical(10), True),
    (Categorical(10), True),
    (Categorical(10), True),
    (Categorical(10), True),
    (Categorical(10), True),
    (Categorical(10), True),
    (Categorical(10), True),
    (Categorical(10), True),
    (Categorical(10), True),
]
c3_celebA_image_size = 32
```

Can you elaborate a bit more in words what you're having problems with? I'm not sure I understand what's not working for you.

**Comment by neocxi** [comment]
> The if network_type == "mnist": line strongly implies that the model would look different for other tasks.

Other experiments were performed using a different (and more messy!) codebase. As we prepare final version of our paper, we expect to port experiments to this codebase.

**Comment by NHDaly** [comment]
> Other experiments were performed using a different (and more messy!) codebase. As we prepare final version of our paper, we expect to port experiments to this codebase.

:) That's amazing, thanks! I've been trying to reproduce the results from the CelebA experiment, but maybe I'll wait until you post your code here. I really appreciate the open science!! ❤️  Thanks!!

**Comment by tachim** [comment]
Any word on the updated experiment code?

**Comment by yunhunJang** [comment]
Has anyone re-implemented celebA experiment results? I've been trying to make it works, but it gives me some noise like outputs. I used model architecture and learning rates described in InfoGAN paper appendix. 
In my experiments, discriminator classifies the real/fake input almost surely from very beginning of the training, and generator does not getting better..

I used same code with mnist experiment except data loading part and model definition (I checked real input image loaded well). Are there any tips or tricks?

Thanks!
